js/ui/windowManager.js

```
@@ -2166,3 +1888,368 @@ const WindowManager = new Lang.Class({
+                                                dst_geometry.x,
+                                                dst_geometry.y,
+                                                dst_geometry.width,
+                                                dst_geometry.height);
+
+
+        /* Tween both windows in a rotation animation at the same time
+        * with backface culling enabled on both. This will allow for
+        * a smooth transition. */
+        Tweener.addTween(src, {
+                        rotation_angle_y: direction == Gtk.DirectionType.RIGHT ? 180 : -180,
+                        time: WINDOW_ANIMATION_TIME * 4,
+                        transition: 'easeOutQuad',
+                        onComplete: Lang.bind(this, function() {
+                            this.rotateOutCompleted(src);
+                        })
+        });
+        Tweener.addTween(dst, {
+                        rotation_angle_y: 0,
+                        time: WINDOW_ANIMATION_TIME * 4,
+                        transition: 'easeOutQuad',
+                        onComplete: Lang.bind(this, function() {
+                            this.rotateInCompleted(dst);
+                        })
+        });
+
+        /* Gently fade the window in, this will paper over
+         * any artifacts from shadows and the like */
+        Tweener.addTween(dst, {
+                    opacity: 255,
+                    time: WINDOW_ANIMATION_TIME * 4,
+                    transition: 'linear'
+        });
+    },
+
+    _removeEffect : function(list, actor) {
+        let idx = list.indexOf(actor);
+        if (idx != -1) {
+            list.splice(idx, 1);
+            return true;
+        }
+        return false;
+    },
+
+    _removeSession : function(session) {
+        let idx = this._sessions.indexOf(session);
+        if (idx != -1) {
+            this._sessions.splice(idx, 1);
+            return true;
+        }
+        return false;
```

Should we return an error?